toolchain/base/value_store.h

```diff
@@ -64,83 +60,56 @@ class ValueStore
                          llvm::StringRef, const ValueType&>;
 
   ValueStore() = default;
-  ValueStore(ValueStore&& other) noexcept
-      : values_((other.UnpoisonAll(), std::move(other.values_)))
-#if CARBON_POISON_VALUE_STORES
-        ,
-        all_poisoned_(false)
-#endif
-  {
-    PoisonAll();
-  }
-  auto operator=(ValueStore&& other) noexcept -> ValueStore& {
-    UnpoisonAll();
-    other.UnpoisonAll();
-    values_ = std::move(other.values_);
-#if CARBON_POISON_VALUE_STORES
-    all_poisoned_ = false;
-#endif
-    PoisonAll();
-    return *this;
-  }
-  ~ValueStore() { UnpoisonAll(); }
 
   // Stores the value and returns an ID to reference it.
   auto Add(ValueType value) -> IdT {
-    IdT id(values_.size());
-    // This routine is especially hot and the check here relatively expensive
-    // for the value provided, so only do this in debug builds to make tracking
-    // down issues easier.
-    CARBON_DCHECK(id.index >= 0, "Id overflow");
-
-    bool realloc = values_.capacity() == values_.size();
-    if (realloc) {
-      // Unpoison everything if the push will reallocate, in order to allow the
-      // vector to make a copy of the elements.
-      UnpoisonAll();
-    } else {
-      PoisonAll();
-    }
-
-    values_.push_back(std::move(value));
-
-    if (realloc) {
-      PoisonAll();
-    } else {
-      PoisonElement(id.index);
+    auto id = IdT(size_);
```

Comment:
Previously we had a check for wrapping (line 91-94 on the delta's left), is that around somewhere? e.g.:

```suggestion
    // This routine is especially hot and the check here relatively expensive
    // for the value provided, so only do this in debug builds to make tracking
    // down issues easier.
    CARBON_DCHECK(size_ < std::numeric_limits<int32_t>::max(), "Id overflow");

    IdT id(size_);
```

(also, very small construction comment slipped in there -- we should avoid the use of assign-to-`auto` here, since a regular construction works, [for style](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md), "Prefer parenthesized initialization (FooType foo(10);) in most other cases.")

Reply:
yes the (d)check is in `IdToChunkIndices()`. It checks for overflow in both index components, which together also fit into an i32. I also mention it in the comment on `size_` below.

Reply:
Sorry, I must be missing how that would work. Let's suppose that size_ going into this function is equal to int32_t max. Can you walk me through what the behavior will be? Won't IdToChunkIndices have a valid input, and size_ will now be invalid?

Reply:
That is, I don't think IdToChunkIndices has invalid input until the *next* call, which may not occur, or size_ may be used before the next call.

Reply:
Ah right no you're right. Will add the wrap around check on size_ here.

Reply:
I think this removes the value of the checks in IdToChunkIndices() then as well, with the contiguous id values that we have again now.